lib/semantics/resolve-names.cc

```diff
@@ -1037,11 +1037,260 @@ class ConstructVisitor : public DeclarationVisitor {
   void PopAssociation();
 };
 
+// Resolve OpenMP construct entities and statement (TODO) entities
+class OmpVisitor : public virtual DeclarationVisitor {
+public:
+  static const parser::Name *GetDesignatorNameIf(
+      const parser::Designator &designator) {
+    const auto *dataRef{std::get_if<parser::DataRef>(&designator.u)};
+    return dataRef ? std::get_if<parser::Name>(&dataRef->u) : nullptr;
+  }
+
+  static constexpr Symbol::Flags dataSharingAttributeFlags{
+      Symbol::Flag::OmpShared, Symbol::Flag::OmpPrivate,
+      Symbol::Flag::OmpFirstPrivate, Symbol::Flag::OmpLastPrivate,
+      Symbol::Flag::OmpReduction, Symbol::Flag::OmpLinear};
+
+  static constexpr Symbol::Flags ompFlagsRequireNewSymbol{
+      Symbol::Flag::OmpPrivate, Symbol::Flag::OmpLinear,
+      Symbol::Flag::OmpFirstPrivate, Symbol::Flag::OmpLastPrivate,
+      Symbol::Flag::OmpReduction};
+
+  static constexpr Symbol::Flags ompFlagsRequireMark{
+      Symbol::Flag::OmpThreadprivate};
+
+  bool Pre(const parser::OpenMPBlockConstruct &) {
+    PushScope(Scope::Kind::Block, nullptr);
+    return true;
+  }
+  void Post(const parser::OpenMPBlockConstruct &) { PopScope(); }
+  bool Pre(const parser::OmpBeginBlockDirective &) {
+    ClearDataSharingAttributeObjects();
+    return true;
+  }
+
+  bool Pre(const parser::OpenMPLoopConstruct &) {
+    PushScope(Scope::Kind::Block, nullptr);
+    return true;
+  }
+  void Post(const parser::OpenMPLoopConstruct &) { PopScope(); }
+  bool Pre(const parser::OmpBeginLoopDirective &) {
+    ClearDataSharingAttributeObjects();
+    return true;
+  }
+
+  bool Pre(const parser::OpenMPThreadprivate &x) {
+    PushScope(Scope::Kind::Block, nullptr);
+    const auto &list{std::get<parser::OmpObjectList>(x.t)};
+    ResolveOmpObjectList(list, Symbol::Flag::OmpThreadprivate);
+    PopScope();
+    return false;
+  }
+
+  bool Pre(const parser::OmpClause::Shared &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpShared);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Private &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Firstprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpFirstPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Lastprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpLastPrivate);
+    return false;
+  }
+
+private:
+  void add_dataSharingAttributeObject(const Symbol *object) {
```

Comment:
Is there a reason for not using Camel case for this name?

Reply:
I did a grep in the `lib/semantics`. Looks like all the `Add` utils are named with `add_` and all the `Clear` utils are named with Camel case.

Reply:
It may or may not be worth fixing those instances in `semantics`, but I don't think that they are a good precedent to follow.  `AddDataSharingAttributeObject` would be better.

Reply:
Thank you both, I will change that.